### ISSUE TITLE
GHCJS spawns a large number of thrunner processes that use a vast amount of memory

### ISSUE THREAD

**Issue by kfigiela** [ghcjs/ghcjs#372]
We build a project of 44 files and we use `lens` extensively, so that GHCJS runs a number of node processes (`thrunner.js`) to process Template Haskell. Each process uses ~130 MB of RAM when idle. At the end of the build total RAM usage for our project is as high as 5 GB (see screenshot) and this becames a big issue and soon may became a blocker.

It seems that there is one node process per file that uses Template Haskell. In case of regular compilation (i.e. by `cabal install`) they could be disposed as soon as TH finishes. In case of `ghcjs --interactive` environment they seem to be reused by GHCJS when the file is changed and reloaded. 

It would be very valuable to optimize that by i.e. running TH in one node process, or maybe they could do better in GC?

![zrzut ekranu 2015-07-21 14 44 59](https://cloud.githubusercontent.com/assets/174118/8800955/50686d2a-2fb7-11e5-84d5-660a569cb57c.png)

**Comment by luite** [comment]
it should be relatively easy to fix this, by closing the streams after a module is done compiling. that still starts a node.js process for each file, but at least the don't all run at the same time.

**Comment by kfigiela** [comment]
We did an experiment and forced GHCJS to reuse single `thrunner` process by setting module name in `getThRunner` to constant value. The project compiles well and now it uses only 130 MB for `thrunner` (one node process). There is also another perk of this workaround: the compilation speeded up.

```haskell
-- TH.sh – from line 242 -- 
getThRunner :: GhcjsEnv -> HscEnv -> DynFlags -> Module -> IO ThRunner
getThRunner js_env hsc_env dflags m = do
  let m' = "DUPA"
  modifyMVar (thRunners js_env) $ \runners -> case M.lookup m' runners of
    Just r  -> return (runners, r)
    Nothing -> do
      r <- startThRunner dflags js_env hsc_env
      -- fixme find where ghc finalizes
      -- when $ TH.qAddModFinalizer (TH.Q $ finishTh is_io js_env m r)
      return (M.insert m' r runners, r)
```

**Comment by luite** [comment]
can you check if the latest commit to `improved-base`, 0f49ed8ed76ae69161e1e8596c06584013e46a1c , fixes this?

Running everything in a single process is risky, since parallel compilation (`-j`) option requires code for multiple modules to be active. We could add something like a process pool, where existing thrunners can get recycled when the module is done, but some code to reset the state needs to be added.

**Comment by kfigiela** [comment]
Yes, indeed this fixes this. However, it kills any speedup we gained by using `ghcjs --interactive`. I suspect that the culprit may long warmup time of `thrunner` process related to loading runtime code. I wonder if we can speed this up. Maybe it would be faster if runtime was read by nodejs from file instead loading it dynamically throught `stdin` as it requires encoding etc from what I see.

**Comment by luite** [comment]
I don't think the encoding is really all that problematic, since it's just base16, handled by built-in node.js functions (since for some reason sending binary data breaks). It could very well be the code loading, the types for TH are pretty big, and it takes a moment for the JS engine to get up to speed, first compiling the code with the baseline JIT compiler, collecting type information, recompiling with the optimizing compiler.

It wouldn't be too hard to add the code for starting a new TH session in the same runner, but there's another tradeoff: There is currently no way to unload code. All symbol names are top level and GHCJS does not track all names (only those visible outside linkable units). So the cost would be increased memory consumption. (this is also one of the reasons that I've marked `ghcjsi` as experimental)

I'd prefer to fix this when the new codegen has landed, since that makes it easier to unload code by just unsetting a variable for a module. Let me know if it's a major problem for you, perhaps we can work something out.

**Comment by kfigiela** [comment]
Thanks, I think we will stick to our ugly fix. In the meantime I will look at https://nodejs.org/api/vm.html, it may have feature that will provide with better workaround before new codegen is ready.

**Comment by qrilka** [comment]
@luite could you please give an update on this issue and the status of that new codegen you were referring too?
